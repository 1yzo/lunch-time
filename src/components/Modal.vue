<template>
    <transition name="fade">
        <div v-if="isVisible" class="modal-mask">
            <div class="modal">
                <i class="fas fa-times close-button" @click="$emit('closeModal')"></i>
                <!-- render coffee message if idList.length === 1 otherwise lunch message and pass optional prop for error -->
                <div v-if="error" class="error">{{ error }}</div>
                <div v-if="names && names.length === 1">Get coffee with {{ names[0] }}</div>
                <div v-if="names && names.length > 1" class="names-container">
                    <div class="name" v-for="name in names" :key="name">{{ name }}</div>
                </div>
            </div>
        </div>
    </transition>
</template>

<script>
export default {
    props: {
        isVisible: {
            type: Boolean,
            required: true
        },
        idList: {
            type: Array,
            required: true
        },
        error: String,
        users: {
            type: Array,
            required: true
        }
    },
    created: function () {
        window.addEventListener('click', (e) => {
            if (e.target.className === "modal-mask") {
                this.$emit('closeModal');
            }
        });

        window.addEventListener('keyup', (e) => {
            if (e.keyCode === 27) {
                if (this.isVisible) {
                    this.$emit('closeModal');
                }
            }
        });
    },
    computed: {
        names: function () {
            return this.users.reduce((acc, curr) => {
                if (this.idList.includes(curr.id)) {
                    return acc.concat(curr.name);
                } else return acc;
            }, []);
        }
    }
};
</script>

<style lang="scss">
.modal-mask {
    position: fixed;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    background-color: rgba(0, 0, 0, .5);
}

.modal {
    position: absolute;
    width: 300px;
    height: 300px;
    background: white;
    box-shadow: 0 1px 4px gray;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    padding: 20px;
}

.close-button {
    cursor: pointer;
    margin-bottom: 20px;
}

.fade-enter-active, .fade-leave-active {
  transition: opacity .5s;
}
.fade-enter, .fade-leave-to {
  opacity: 0;
}
</style>