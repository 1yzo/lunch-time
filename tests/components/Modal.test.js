import Vue from 'vue';
import { shallowMount } from '@vue/test-utils';
import Modal from '../../src/components/Modal';
import { coffeeUsers } from '../fixtures/users';

const createWrapper = propsData => shallowMount(Modal, { propsData });

describe('Modal', () => {
    it('renders nothing if isVisible prop is false', () => {
        const wrapper = createWrapper({
            isVisible: false,
            users: coffeeUsers,
            idList: []
        });
        expect(wrapper.element).toMatchSnapshot();
    })
    
    it('renders coffee partner if idList contains only 1 id', () => {
        const wrapper = createWrapper({
            isVisible: true,
            idList: [coffeeUsers[0].id],
            users: coffeeUsers
        });
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders lunch group if idList contains multiple ids', () => {
        const wrapper = createWrapper({
            isVisible: true,
            idList: coffeeUsers.slice(0, 4).map((user) => user.id),
            users: coffeeUsers,
            currentUserName: 'Current User'
        });
        expect(wrapper.element).toMatchSnapshot();
    });

    it('contains error message if error exists', () => {
        const wrapper = createWrapper({
            isVisible: true,
            idList: [],
            users: coffeeUsers,
            error: 'Something went wrong'
        });
        expect(wrapper.contains('.error')).toBeTruthy();
    });

    it('correctly computes names from idList', () => {
        const wrapper = createWrapper({
            isVisible: true,
            idList: coffeeUsers.slice(0, 4).map((user) => user.id),
            users: coffeeUsers
        });
        expect(wrapper.vm.names).toEqual(['first', 'second', 'third', 'fourth']);
    });

    it('emits closeModal on close button click', () => {
        const wrapper = createWrapper({
            isVisible: true,
            idList: coffeeUsers.slice(0, 4).map((user) => user.id),
            users: coffeeUsers
        });
        wrapper.find('.close-button').trigger('click');
        expect(wrapper.emitted()).toHaveProperty('closeModal');
    });

    it('emits closeModal on modal-mask click', () => {
        const wrapper = createWrapper({
            isVisible: true,
            idList: coffeeUsers.slice(0, 4).map((user) => user.id),
            users: coffeeUsers
        });
        wrapper.find('.modal-mask').trigger('click');
        expect(wrapper.emitted()).toHaveProperty('closeModal');
    });
});