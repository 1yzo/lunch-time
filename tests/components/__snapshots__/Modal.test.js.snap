// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders coffee partner if idList contains only 1 id 1`] = `
<div
  class="modal-mask"
  name="fade"
>
  <div
    class="modal"
  >
    <i
      class="fas fa-times close-button"
    />
     
    <!---->
     
    <div
      class="title names-container"
    >
      <div>
        Get coffee with
      </div>
       
      <div
        class="name"
      >
        first
      </div>
    </div>
     
    <!---->
  </div>
</div>
`;

exports[`Modal renders lunch group if idList contains multiple ids 1`] = `
<div
  class="modal-mask"
  name="fade"
>
  <div
    class="modal"
  >
    <i
      class="fas fa-times close-button"
    />
     
    <!---->
     
    <!---->
     
    <div>
      <div
        class="title"
      >
        Today's lunch group
      </div>
       
      <div
        class="names-container"
      >
        <div
          class="name"
        >
          Current User
        </div>
         
        <div
          class="name"
        >
          first
        </div>
        <div
          class="name"
        >
          second
        </div>
        <div
          class="name"
        >
          third
        </div>
        <div
          class="name"
        >
          fourth
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal renders nothing if isVisible prop is false 1`] = `<!---->`;
